File: Pinball App 2/Pinball App 2/gameroom/GameRoomCatalogLoader.swift
import Foundation
import Combine
import OSLog

let gameRoomCatalogLogger = Logger(
    subsystem: Bundle.main.bundleIdentifier ?? "com.pillyliu.Pinball-App-2",
    category: "DataIntegrity"
)

@MainActor
final class GameRoomCatalogLoader: ObservableObject {
    @Published private(set) var games: [GameRoomCatalogGame] = []
    @Published private(set) var variantOptionsByCatalogGameID: [String: [String]] = [:]
    @Published private(set) var isLoading = false
    @Published private(set) var errorMessage: String?

    private var didLoad = false
    private var allCatalogGames: [GameRoomCatalogGame] = []
    private var gamesByCatalogGameID: [String: [GameRoomCatalogGame]] = [:]
    private var gamesByNormalizedCatalogGameID: [String: [GameRoomCatalogGame]] = [:]
    private var variantOptionsByNormalizedCatalogGameID: [String: [String]] = [:]
    private var slugMatchesByKey: [String: GameRoomCatalogSlugMatch] = [:]

    func loadIfNeeded() async {
        guard !didLoad else { return }
        didLoad = true
        await reload()
    }

    func reload() async {
        isLoading = true
        errorMessage = nil

        do {
            games = try await loadCatalogGames()
        } catch {
            errorMessage = "Failed to load catalog data: \(error.localizedDescription)"
        }

        isLoading = false
    }

    private func loadCatalogGames() async throws -> [GameRoomCatalogGame] {
        let practiceIdentityCurationsData = try await loadHostedOrCachedPinballJSONData(
            path: hostedPracticeIdentityCurationsPath,
            allowMissing: true
        )
        guard let rawData = try await loadHostedOrCachedPinballJSONData(
            path: hostedOPDBExportPath,
            allowMissing: true
        ),
        let rawMachines = try? decodeOPDBExportCatalogMachines(
            data: rawData,
            practiceIdentityCurationsData: practiceIdentityCurationsData
        ),
        !rawMachines.isEmpty else {
            throw URLError(.resourceUnavailable)
        }

        let mappedGames = rawMachines.map(Self.makeGame)
        allCatalogGames = mappedGames
        gamesByCatalogGameID = Dictionary(grouping: mappedGames, by: \.catalogGameID)
        gamesByNormalizedCatalogGameID = Dictionary(grouping: mappedGames, by: { Self.normalizedCatalogGameID($0.catalogGameID) })
        variantOptionsByCatalogGameID = Self.variantOptionsMap(from: rawMachines)
        variantOptionsByNormalizedCatalogGameID = variantOptionsByCatalogGameID.reduce(into: [:]) { partialResult, entry in
            partialResult[Self.normalizedCatalogGameID(entry.key)] = entry.value
        }
        slugMatchesByKey = Self.slugMatches(from: rawMachines)
        return Self.dedupedGames(from: mappedGames)
    }

    func variantOptions(for catalogGameID: String) -> [String] {
        variantOptionsByCatalogGameID[catalogGameID] ??
            variantOptionsByNormalizedCatalogGameID[Self.normalizedCatalogGameID(catalogGameID)] ??
            []
    }

    func games(for catalogGameID: String) -> [GameRoomCatalogGame] {
        if let grouped = gamesByCatalogGameID[catalogGameID], !grouped.isEmpty {
            return grouped.sorted(by: Self.sortGames)
        }
        if let grouped = gamesByNormalizedCatalogGameID[Self.normalizedCatalogGameID(catalogGameID)], !grouped.isEmpty {
            return grouped.sorted(by: Self.sortGames)
        }
        return []
    }

    func game(for catalogGameID: String) -> GameRoomCatalogGame? {
        if let grouped = gamesByCatalogGameID[catalogGameID], !grouped.isEmpty {
            return Self.preferredGame(in: grouped)
        }
        if let grouped = gamesByNormalizedCatalogGameID[Self.normalizedCatalogGameID(catalogGameID)], !grouped.isEmpty {
            return Self.preferredGame(in: grouped)
        }
        return games.first(where: { $0.catalogGameID.caseInsensitiveCompare(catalogGameID) == .orderedSame })
    }

    func game(for catalogGameID: String, variant: String?) -> GameRoomCatalogGame? {
        if let normalizedVariant = Self.normalizedVariant(variant) {
            let grouped = games(for: catalogGameID)
            let exactMatches = grouped.filter {
                Self.exactVariantMatchesSelection(candidate: $0.displayVariant, selected: normalizedVariant)
            }
            if !exactMatches.isEmpty {
                return Self.preferredGame(in: exactMatches)
            }
            let matches = grouped.filter {
                Self.variantMatchesSelection(candidate: $0.displayVariant, selected: variant)
            }
            if !matches.isEmpty {
                return Self.preferredGame(in: matches)
            }
        }
        return game(for: catalogGameID)
    }

    func slugMatch(for slug: String) -> GameRoomCatalogSlugMatch? {
        Self.buildSlugKeys(from: slug).first { slugMatchesByKey[$0] != nil }.flatMap { slugMatchesByKey[$0] }
    }

    func resolvedOPDBID(for machine: OwnedMachine) -> String? {
        gameRoomResolvedOPDBID(
            for: machine,
            allCatalogGames: allCatalogGames,
            gamesByCatalogGameID: gamesByCatalogGameID,
            gamesByNormalizedCatalogGameID: gamesByNormalizedCatalogGameID
        )
    }

    func normalizedCatalogGame(for machine: OwnedMachine) -> GameRoomCatalogGame? {
        gameRoomNormalizedCatalogGame(
            for: machine,
            allCatalogGames: allCatalogGames,
            gamesByCatalogGameID: gamesByCatalogGameID,
            gamesByNormalizedCatalogGameID: gamesByNormalizedCatalogGameID
        )
    }

    func imageCandidates(for machine: OwnedMachine) -> [URL] {
        gameRoomCatalogImageCandidates(
            for: machine,
            allCatalogGames: allCatalogGames,
            gamesByCatalogGameID: gamesByCatalogGameID,
            gamesByNormalizedCatalogGameID: gamesByNormalizedCatalogGameID
        )
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/gameroom/GameRoomCatalogLoader.kt
package com.pillyliu.pinprofandroid.gameroom

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.pillyliu.pinprofandroid.data.PinballDataCache
import com.pillyliu.pinprofandroid.library.hostedOPDBExportPath
import com.pillyliu.pinprofandroid.library.hostedPracticeIdentityCurationsPath
import kotlinx.coroutines.runBlocking

internal class GameRoomCatalogLoader {
    var didLoad by mutableStateOf(false)
        private set

    var isLoading by mutableStateOf(false)
        private set

    var errorMessage by mutableStateOf<String?>(null)
        private set

    var games by mutableStateOf<List<GameRoomCatalogGame>>(emptyList())
        private set

    var manufacturers by mutableStateOf<List<String>>(emptyList())
        private set

    var manufacturerOptions by mutableStateOf<List<GameRoomCatalogManufacturerOption>>(emptyList())
        private set

    var variantOptionsByCatalogGameID by mutableStateOf<Map<String, List<String>>>(emptyMap())
        private set

    private var allCatalogGames: List<GameRoomCatalogGame> = emptyList()
    private var gamesByCatalogGameID: Map<String, List<GameRoomCatalogGame>> = emptyMap()
    private var gamesByNormalizedCatalogGameID: Map<String, List<GameRoomCatalogGame>> = emptyMap()
    private var variantOptionsByNormalizedCatalogGameID: Map<String, List<String>> = emptyMap()
    private var machineRecordsByCatalogGameID: Map<String, List<GameRoomCatalogMachineRecord>> = emptyMap()
    private var slugMatchesBySlug: Map<String, GameRoomCatalogSlugMatch> = emptyMap()

    private val resolutionContext: GameRoomCatalogResolutionContext
        get() = GameRoomCatalogResolutionContext(
            allCatalogGames = allCatalogGames,
            gamesByCatalogGameID = gamesByCatalogGameID,
            gamesByNormalizedCatalogGameID = gamesByNormalizedCatalogGameID,
            machineRecordsByCatalogGameID = machineRecordsByCatalogGameID,
        )

    fun loadIfNeeded() {
        if (didLoad) return
        didLoad = true
        isLoading = true
        errorMessage = null
        try {
            loadCatalog()
        } catch (error: Throwable) {
            games = emptyList()
            manufacturers = emptyList()
            manufacturerOptions = emptyList()
            variantOptionsByCatalogGameID = emptyMap()
            allCatalogGames = emptyList()
            gamesByCatalogGameID = emptyMap()
            gamesByNormalizedCatalogGameID = emptyMap()
            variantOptionsByNormalizedCatalogGameID = emptyMap()
            machineRecordsByCatalogGameID = emptyMap()
            slugMatchesBySlug = emptyMap()
            errorMessage = "Failed to load catalog data: ${error.localizedMessage ?: error::class.java.simpleName}"
        }
        isLoading = false
    }

    private fun loadCatalog() {
        val raw = runBlocking {
            PinballDataCache.loadText(
                url = hostedOPDBExportPath,
                allowMissing = false,
            ).text
        } ?: throw IllegalStateException("Catalog data is missing.")
        val practiceIdentityCurationsRaw = runBlocking {
            PinballDataCache.loadText(
                url = hostedPracticeIdentityCurationsPath,
                allowMissing = true,
            ).text
        }?.takeIf { it.isNotBlank() }
        val loadedCatalog = buildGameRoomLoadedCatalogData(raw, practiceIdentityCurationsRaw)
        allCatalogGames = loadedCatalog.allCatalogGames
        games = loadedCatalog.games
        manufacturers = loadedCatalog.manufacturers
        manufacturerOptions = loadedCatalog.manufacturerOptions
        gamesByCatalogGameID = loadedCatalog.gamesByCatalogGameID
        gamesByNormalizedCatalogGameID = loadedCatalog.gamesByNormalizedCatalogGameID
        variantOptionsByCatalogGameID = loadedCatalog.variantOptionsByCatalogGameID
        variantOptionsByNormalizedCatalogGameID = loadedCatalog.variantOptionsByNormalizedCatalogGameID
        machineRecordsByCatalogGameID = loadedCatalog.machineRecordsByCatalogGameID
        slugMatchesBySlug = loadedCatalog.slugMatchesBySlug
        errorMessage = null
    }

    fun variantOptions(catalogGameID: String): List<String> {
        variantOptionsByCatalogGameID[catalogGameID]
            ?.let { return it }
        return variantOptionsByNormalizedCatalogGameID[normalizedCatalogGameID(catalogGameID)].orEmpty()
    }

    fun game(catalogGameID: String): GameRoomCatalogGame? {
        val normalizedID = catalogGameID.trim().takeIf { it.isNotEmpty() } ?: return null
        gamesByCatalogGameID[normalizedID]
            ?.let(::preferredCatalogGame)
            ?.let { return it }
        gamesByNormalizedCatalogGameID[normalizedCatalogGameID(normalizedID)]
            ?.let(::preferredCatalogGame)
            ?.let { return it }
        return games.firstOrNull { it.catalogGameID.equals(normalizedID, ignoreCase = true) }
    }

    fun games(catalogGameID: String): List<GameRoomCatalogGame> {
        val normalizedID = catalogGameID.trim().takeIf { it.isNotEmpty() } ?: return emptyList()
        gamesByCatalogGameID[normalizedID]?.let { return it.sortedWith(::compareSortedCatalogGames) }
        gamesByNormalizedCatalogGameID[normalizedCatalogGameID(normalizedID)]?.let { return it.sortedWith(::compareSortedCatalogGames) }
        return emptyList()
    }

    fun game(catalogGameID: String, variant: String?): GameRoomCatalogGame? {
        val normalizedVariant = normalizeVariantLabel(variant)
        if (normalizedVariant != null) {
            val grouped = games(catalogGameID)
            val exactMatches = grouped.filter {
                exactVariantMatchesSelection(it.displayVariant, normalizedVariant)
            }
            if (exactMatches.isNotEmpty()) {
                return preferredCatalogGame(exactMatches)
            }
            val matches = grouped.filter {
                variantMatchesSelection(it.displayVariant, variant)
            }
            if (matches.isNotEmpty()) {
                return preferredCatalogGame(matches)
            }
        }
        return game(catalogGameID)
    }

    fun slugMatch(slug: String): GameRoomCatalogSlugMatch? {
        val normalizedSlug = slug.trim().lowercase()
        if (normalizedSlug.isBlank()) return null
        return buildSlugKeys(normalizedSlug).firstNotNullOfOrNull { key -> slugMatchesBySlug[key] }
    }

    fun resolvedOpdbId(machine: OwnedMachine): String? {
        val grouped = games(machine.catalogGameID)
        return resolvedGameRoomCatalogOpdbId(
            context = resolutionContext,
            machine = machine,
            groupedGames = grouped,
        )
    }

    fun normalizedCatalogGame(machine: OwnedMachine): GameRoomCatalogGame? {
        val exact = resolvedOpdbId(machine) ?: return null
        return gameRoomCatalogGameForExactOpdbId(resolutionContext, exact)
    }

    fun imageCandidates(machine: OwnedMachine): List<String> {
        return gameRoomCatalogImageCandidates(
            context = resolutionContext,
            machine = machine,
            resolvedOpdbId = resolvedOpdbId(machine),
        )
    }

    fun resolvedArt(
        catalogGameID: String,
        opdbID: String? = null,
        selectedVariant: String?,
        selectedTitle: String? = null,
        selectedYear: Int? = null,
    ): GameRoomCatalogArt? {
        return resolveGameRoomCatalogArt(
            context = resolutionContext,
            catalogGameID = catalogGameID,
            opdbID = opdbID,
            selectedVariant = selectedVariant,
            selectedTitle = selectedTitle,
            selectedYear = selectedYear,
        )
    }

}
